framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/log/InstancesLogCollectorRunnable.java

```diff
@@ -0,0 +1,118 @@
+package org.kie.cloud.openshift.log;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.kie.cloud.api.deployment.Instance;
+import org.kie.cloud.common.logs.InstanceLogUtil;
+import org.kie.cloud.openshift.deployment.OpenShiftInstance;
+import org.kie.cloud.openshift.resource.Project;
+import org.kie.cloud.openshift.scenario.OpenShiftScenario;
+import org.kie.cloud.openshift.util.OpenshiftInstanceUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InstancesLogCollectorRunnable implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftScenario.class);
+
+    private static final Integer DEFAULT_OBERVABLE_BUFFER_IN_SECONDS = 5;
+
+    private Project project;
+    private String logFolderName;
+
+    private Map<OpenShiftInstance, Future<?>> observedInstances = new ConcurrentHashMap<>();
+
+    public InstancesLogCollectorRunnable(Project project, String logFolderName) {
+        super();
+        this.project = project;
+        this.logFolderName = logFolderName;
+    }
+
+    @Override
+    public void run() {
+        // Check for new instances and oberve on them
+        OpenshiftInstanceUtil.getAllInstances(project.getOpenShift(), project.getName()).stream()
+                             // Filter non observed instances
+                             .filter(instance -> !isInstanceObserved(instance))
+                             // Observe instance logs
+                             .forEach(this::observeInstanceLog);
+
+        // Check if instances are still existing
+        // Calling get pod on a no more existing pod will cause the "obervePodLog" to complete ...
+    }
+
+    public void closeAndFlushRemainingInstanceCollectors() {
+        // For all remaining observed instances, stop thread && writeinstancelogs
+        List<OpenShiftInstance> instances = new ArrayList<>(observedInstances.keySet());
+        instances.forEach(this::stopObserveInstanceLogs);
+        instances.forEach(this::flushInstanceLogs);
+    }
+
+    private void observeInstanceLog(OpenShiftInstance instance) {
+        Future<?> future = Executors.newSingleThreadExecutor().submit(() -> {
```

Comment:
Unnecessarily creates a new ExecutorService instance for each pod. Let's reuse a single instance and give it as many threads as many pods we have.

Reply:
modified with executor